### PR TITLE
Clarify return value of function in fourier.coefficients

### LINKED ISSUE #1449: `qml.fourier.coefficients` won't  work with more than one expectation value
#### Issue description

When I try to calculate the Fourier coefficients with the new module `qml.fourier` if the circuit has more than one measurement an error appears.

#### Code:
```python
import pennylane as qml
from pennylane import numpy as np
from pennylane.fourier import coefficients

dev = qml.device('default.qubit', wires=2)

@qml.qnode(dev)
def simple_circuit(x):
    qml.RX(x[0], wires=0)
    qml.RY(x[0], wires=1)
    qml.CNOT(wires=[1, 0])
    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))


coeffs = coefficients(simple_circuit, 1, 2)
print(np.round(coeffs, decimals=4))
```
#### Output:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
TypeError: only size-1 arrays can be converted to Python scalars

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-111-8c2ae78fdea3> in <module>
     13 
     14 
---> 15 coeffs = coefficients(simple_circuit, 1, 2)
     16 print(np.round(coeffs, decimals=4))

~/.local/lib/python3.8/site-packages/pennylane/fourier/coefficients.py in coefficients(f, n_inputs, degree, lowpass_filter, filter_threshold)
    154     """
    155     if not lowpass_filter:
--> 156         return _coefficients_no_filter(f, n_inputs, degree)
    157 
    158     if filter_threshold is None:

~/.local/lib/python3.8/site-packages/pennylane/fourier/coefficients.py in _coefficients_no_filter(f, n_inputs, degree)
    228 
    229         # fill discretized function array with value of f at inpts
--> 230         f_discrete[tuple(nvec)] = f(sample_points)
    231 
    232     coeffs = np.fft.fftn(f_discrete) / f_discrete.size

ValueError: setting an array element with a sequence.
```

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context**

The docs are currently not very clear on what kinds of qnodes the `fourier.coefficients` function can be applied to. Adding a sentence to clarify.

**Related issue**

Closes #1449.

### CODE REVIEW

File: pennylane/fourier/coefficients.py [hunk 1]
```diff
@@ -55,7 +55,8 @@ def coefficients(f, n_inputs, degree, lowpass_filter=False, filter_threshold=Non
     :math:`c_{n_1,\ldots,n_N}` are Fourier coefficients.
 
     Args:
-        f (callable): Function that takes a 1D tensor of ``n_inputs`` scalar inputs.
+        f (callable): Function that takes a 1D tensor of ``n_inputs`` scalar inputs. The function can be a QNode, but
+            has to return a single real value (such as an expectation) only.
```
Comment (mariaschuld) [inline pennylane/fourier/coefficients.py:59]:
```suggestion
            has to return a real scalar value (such as an expectation).
```

File: pennylane/fourier/coefficients.py [hunk 2]
```diff
@@ -55,7 +55,8 @@ def coefficients(f, n_inputs, degree, lowpass_filter=False, filter_threshold=Non
     :math:`c_{n_1,\ldots,n_N}` are Fourier coefficients.
 
     Args:
-        f (callable): Function that takes a 1D tensor of ``n_inputs`` scalar inputs.
+        f (callable): Function that takes a 1D tensor of ``n_inputs`` scalar inputs. The function can be a QNode, but
+            has to return a real scalar value (such as an expectation).
```
Comment (antalszava) [inline pennylane/fourier/coefficients.py:59]: Just to clarify, it would not work with return types such as `qml.probs`, `qml.state`, etc. correct?

**Review by antalszava — COMMENTED** [review]
Thank you @mariaschuld! It would also be great to have error handling for this so that the users get a descriptive message. At the moment the error is
```python
ValueError: setting an array element with a sequence.
```

File: pennylane/fourier/coefficients.py (hunk 2 above)
Comment (mariaschuld) [reply pennylane/fourier/coefficients.py:59]: Yes!

**Comment by mariaschuld** [comment]
> Thank you @mariaschuld! It would also be great to have error handling for this so that the users get a descriptive message. At the moment the error is
> 
> ```python
> ValueError: setting an array element with a sequence.
> ```

Hmm... this will not be so easy in fact. We are passing the function to a scipy function without evaluating it. I fear the onus has to be on the user here, unless we want to hack something in. Or do you have an idea @glassnotes ?

**Comment by glassnotes** [comment]
> Hmm... this will not be so easy in fact. We are passing the function to a scipy function without evaluating it. I fear the onus has to be on the user here, unless we want to hack something in. Or do you have an idea @glassnotes ?

@mariaschuld what if we just do a basic check at the beginning of `coefficients` that executes the function once (or `try`/`except`) and inspects the return value?

**Comment by josh146** [comment]
> what if we just do a basic check at the beginning of coefficients that executes the function once (or try/except) and inspects the return value?

Could this result in unexpected quantum evaluations? I have to admit, I lean towards not doing this check in that case, and instead strongly documenting it.

**Comment by glassnotes** [comment]
> Could this result in unexpected quantum evaluations? I have to admit, I lean towards not doing this check in that case, and instead strongly documenting it.

Oh that's a very good point, I didn't think of that! Yeah maybe just documentation is the way to go then (we could bold or otherwise emphasize that restriction in the docstring).

**Comment by antalszava** [comment]
If the function that was passed is a QNode, then we could check the number of returns of the qfunc. This could be done just by calling it just as we do when queuing.

```python
In [6]: qnode1
Out[6]: <QNode: wires=3, device='default.qubit.autograd', interface='autograd', diff_method='best'>

In [9]: qnode1.func
Out[9]: <function __main__.circuit()>

In [10]: qnode1.func()
Out[10]: 
(expval(PauliZ(wires=[0])),
 expval(PauliZ(wires=[1])),
 expval(PauliZ(wires=[2])))
```
Users will likely use the function for QNodes, so it might make sense to check this, otherwise the documentation should work well.

**Comment by mariaschuld** [comment]
> If the function that was passed is a QNode, then we could check the number of returns of the qfunc. This could be done just by calling it as we do when queuing:
> 
> ```python
> In [6]: qnode1
> Out[6]: <QNode: wires=3, device='default.qubit.autograd', interface='autograd', diff_method='best'>
> 
> In [9]: qnode1.func
> Out[9]: <function __main__.circuit()>
> 
> In [10]: qnode1.func()
> Out[10]: 
> (expval(PauliZ(wires=[0])),
>  expval(PauliZ(wires=[1])),
>  expval(PauliZ(wires=[2])))
> ```
> 
> Users will likely use `coefficients` with QNodes, so it might make sense to check this, otherwise, the documentation should work well.

Hmm, it's quite a fundamental questions: if we provide functions that take a function f and wrap it, do we need to check that the user constructed f with the right specs? I tend to say no here, as long as it's in the docstring. I would therefore vote to merge, and if we get another user being confused from here we add an exception. 

But happy to be outvoted and add the check :)

**Review by antalszava — APPROVED** [review]
:100: 

Could we add another note in the examples section (e.g., on line 90)? E.g.,
 
        .. note::

            The QNode has to return a scalar value (such as an expectation).

**Comment by antalszava** [comment]
[ch6962]

**Comment by mariaschuld** [comment]
Added the note, thanks for the review!

**Pull request merged**